### DISCUSSION TITLE
New in 1.30 with "-bash: !: event not found" when trying to Manually fixing data

### DISCUSSION THREAD

**Discussion by GtHbbrr** [teslamate-org/teslamate#4065] [Q&A] [answered]
I noticed the list of incomplete drives:
From here you can check if you have incomplete data of Drives (drives without ending date) If so, you may follow the official guide by [Manually fixing data](https://docs.teslamate.org/docs/maintenance/manually_fixing_data)

On following the instruction for manually fixing data with replacing the 9999 for severakl of the incomplete drives I get this response:
`docker compose exec teslamate bin/teslamate rpc \
    "TeslaMate.Repo.get!(TeslaMate.Log.Drive, 5029) |> TeslaMate.Log.close_drive()"`

`-bash: !: event not found`

**Comment by djpalmis** [comment]
https://github.com/teslamate-org/teslamate/issues/2524 still this issue claims victims

**Reply by GtHbbrr** [reply]
Any known solutions?

**Reply by DrMichael** [reply]
Replace the " with '

**Reply by JakobLichterfeld** [reply]
> Any known solutions?

For sure, just either make sure copy paste does not break encoding or type the command. Or make sure your Drive id really exist. You can use the drives dashboard to see unfinished drives at the bottom since v1.30.0

**Reply by DrMichael** [reply]
I think, that is a shell issue...

**Reply by GtHbbrr** [reply]
Tried the suggested manual fix after replacing the " with ' but get the same result:
 `-bash: !: event not found`

Why not add some code to close these open drives automatically? You obviously do have the data to fix it

**Reply by GtHbbrr** [reply]
If there is any danger of messing up the database then put in a dialog before fixing or add a setting to do or do not automagically close open drives and or charges

**Reply by GtHbbrr** [reply]
Please take part in the discussion

**Reply by brianmay** [reply]
teslamate is a daemon that runs in the background, it has no opportunity to ask the user questions on startup.

What we could do:

* Automatically close open drives/charges on startup. I imagine the issue here is that the car could currently be driving or charging. Is that going to cause problems? Maybe it is the same charge/drive, in which case we don't want to close the current one. Or maybe it is a new charge/drive in which case we do want to close the current one.

* Add details to the web interface that shows open drives or charges, and give the option to manually close them.

I am kind of inclined to think maybe we should automatically close drives/charges on startup. Or maybe add some sort of threshold  (if this charge/drive hasn't been updated in the last hour close it). Because it is likely that drive/charge has probably ended, and even if the car is driving/charging, it could be a new drive/charge now.

But I am not really familiar with this code, so not really sure if this might break/upset something I haven't thought of.

Somebody needs to come up with a pull request to try to fix this, then we can discuss further if this is a good idea or not.

**Reply by JakobLichterfeld** [reply]
> Please take part in the discussion

I do not have more to add than the solution I added above, which you obviously did not try nor provided additional info.

**Reply by GtHbbrr** [reply]
Before making the effort to start this discussion I made sure copy did not break the command and also typed the instruction. Thank you.

**Reply by GtHbbrr** [reply]
Is there more information available on this command to manually fix drives/charges? I can not find any feed back in the logs

**Reply by GtHbbrr** [reply]
Closing automatically is always ok for drives that started say more than 1 day ago. No single drive in any ev will last longer than 24 hours

**Reply by GtHbbrr** [reply]
Is it not the intention of the teslamate settings page to ask the user questions on startup?

**Reply by GtHbbrr** [reply, marked as answer]
On investigating further following the previous issue #2524 I managed to resolve using the "close multiple drives at once"-script in issue #2364. This script as proposed by [init6-git](https://github.com/init6-git) and further developed by [aiden-cat](https://github.com/aiden-cat), [dmachadol](https://github.com/dmachadol) and [cwanja](https://github.com/cwanja).

Also [cwanja](https://github.com/cwanja) did propose a pull request in PR https://github.com/teslamate-org/teslamate/pull/3877 but this was denied citing:

- The docu is not idemponent
- the script use sudo docker ...
- the bash style is not system neutral
- mentioning a single person for credit

What is idemponent?

**Reply by GtHbbrr** [reply]
I suggest adding a reference to issue #2364 in the documentation for Manually fixing data

**Reply by JakobLichterfeld** [reply]
> Closing automatically is always ok for drives that started say more than 1 day ago. No single drive in any ev will last longer than 24 hours

That's technically not correct.

**Reply by JakobLichterfeld** [reply]
> What is idemponent?

"In computing, an idempotent operation is one that has no additional effect if it is called more than once with the same input parameters."

**Reply by JakobLichterfeld** [reply]
> I managed to resolve using the "close multiple drives at once"-script in issue

Glad you solved your issue. It proves that you do have an issue with your shell and your event name as provided as answer multiple times before.

**Reply by GtHbbrr** [reply]
My resolve just proved the script from issue #2364 to close multiple drives at once does work. I still have no clue how I should manually fix data with the instructions in the documentation.

**Comment by VinxItak** [comment]
Hello, just replace the double quotes by simple quote :
`bin/teslamate rpc "TeslaMate.Repo.get!(TeslaMate.Log.ChargingProcess, 9999) |> TeslaMate.Log.complete_charging_process()"`
by
`bin/teslamate rpc 'TeslaMate.Repo.get!(TeslaMate.Log.ChargingProcess, 9999) |> TeslaMate.Log.complete_charging_process()'`

**Comment by brianmay** [comment]
Wonder if this is specific to older versions of bash maybe. "!" is a special character for history, but my version of bash seems to be OK with that command line.

```
[root@iot2:~]# bash --version
GNU bash, version 5.3.9(1)-release (x86_64-pc-linux-gnu)
Copyright (C) 2025 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

[root@iot2:~]# echo "TeslaMate.Repo.get!(TeslaMate.Log.Drive, 9999)"
TeslaMate.Repo.get!(TeslaMate.Log.Drive, 9999)
```

Just to prove I can break it:

```
[root@iot2:~]# echo "TeslaMate.Repo.get!aa(TeslaMate.Log.Drive, 9999)"
-bash: !aa: event not found
```

According to AI, the following settings could have an impact too (they didn't for me):

```
[root@iot2:~]# shopt extglob
extglob             	on

[root@iot2:~]# shopt -o histexpand
histexpand          	on
```
